bakery/checker.go

```diff
@@ -35,57 +27,82 @@ type Op struct {
 	Action string
 }
 
+// NoOp holds the empty operation, signifying no authorized
+// operation. This is always considered to be authorized.
+var NoOp = Op{}
+
 // CheckerParams holds parameters for NewChecker.
 type CheckerParams struct {
-	// CaveatChecker is used to check first party caveats when authorizing.
+	// Checker is used to check first party caveats when authorizing.
 	// If this is nil NewChecker will use checkers.New(nil).
 	Checker FirstPartyCaveatChecker
 
-	// Authorizer is used to check whether an authenticated user is
-	// allowed to perform operations. If it is nil, NewChecker will
-	// use ClosedAuthorizer.
-	//
-	// The identity parameter passed to Authorizer.Allow will
-	// always have been obtained from a call to
-	// IdentityClient.DeclaredIdentity.
-	Authorizer Authorizer
-
 	// OpsAuthorizer is used to check whether operations are authorized
 	// by some other already-authorized operation. If it is nil,
 	// NewChecker will assume no operation is authorized by any
 	// operation except itself.
 	OpsAuthorizer OpsAuthorizer
 
-	// IdentityClient is used for interactions with the external
-	// identity service used for authentication.
-	//
-	// If this is nil, no authentication will be possible.
-	IdentityClient IdentityClient
-
 	// MacaroonVerifier is used to verify macaroons.
 	MacaroonVerifier MacaroonVerifier
 }
 
+// OpsAuthorizer is used to check whether an operation authorizes some other
+// operation. For example, a macaroon with an operation allowing general access to a service
+// might also grant access to a more specific operation.
+type OpsAuthorizer interface {
+	// AuthorizeOp reports which elements of queryOps are authorized by
+	// authorizedOp. On return, each element of the slice should represent
+	// whether the respective element in queryOps has been authorized.
+	// An empty returned slice indicates that no operations are authorized.
+	// AuthorizeOps may also return third party caveats that apply to
+	// the authorized operations. Access will only be authorized when
+	// those caveats are discharged by the client.
+	//
+	// When not all operations can be authorized with the macaroons
+	// supplied to Checker.Auth, the checker will call AuthorizeOps
+	// with NoOp, because some operations might be authorized
+	// regardless of authority. NoOp will always be the last
+	// operation queried within any given Allow call.
+	AuthorizeOps(ctx context.Context, authorizedOp Op, queryOps []Op) ([]bool, []checkers.Caveat, error)
+}
+
 // AuthInfo information about an authorization decision.
 type AuthInfo struct {
```

Comment:
hmmm.. i can see how i would want to pass a logger to BakeryParams and have all authorization decisions automatically logged at some logging level (say Info)...

Reply:
yeah, that would be nice. I think that's something that can be added later without backward incompatibility though.